### PR TITLE
Comments and styling for ChainBase

### PR DESCRIPTION
**Pull request by mikedennis**
Also removed a confusing implementation of Contains which was not referenced anywhere. See #659 for details.

### CODE REVIEW

File: src/NBitcoin/ChainBase.cs [hunk 1]
```diff
@@ -4,15 +4,33 @@
 
 namespace NBitcoin
 {
+    /// <summary>Base implementation for chains.</summary>
     public abstract class ChainBase
     {
+        /// <summary>
+        /// Gets the block given a block id(hash).
```
Comment (Aprogiena) [inline src/NBitcoin/ChainBase.cs]: i'd wrote ID as with capital I and D, I'd also put space between ID and `(`

but most importantly, here and in whole file, I think we should distinguish between Block and ChainedBlock
as Block is what represents and actual block, but ChainedBlock is imho misnamed, it is actually ChainedBlockHeader, or ChainedHeader, it is not a block 

So if we want to to keep this little confusing name ChainedBlock, we should at least make comments that make it clear that this is not a block. GetBlock is again misnamed method, it does not return a block

GetBlock should be named GetChainedBlock (optimally GetChainedBlockHeader ...) 

So I was wondering if we could at least start with comments being correct

File: src/NBitcoin/ChainBase.cs [hunk 2]
```diff
@@ -4,15 +4,33 @@
...
+        /// </summary>
+        /// <param name="id">Block id to retreive.</param>
+        /// <returns>The block.</returns>
         public abstract ChainedBlock GetBlock(uint256 id);
+
+        /// <summary>
+        /// Gets the block at a given block height.
+        /// </summary>
+        /// <param name="height">Height to retreive block at.</param>
```
Comment (Aprogiena) [inline src/NBitcoin/ChainBase.cs]: typo `ei`, i'd rephrase `Height of the block which header to retrieve.`

File: src/NBitcoin/ChainBase.cs [hunk 3]
```diff
@@ -4,15 +4,33 @@
...
         public abstract ChainedBlock Tip { get; }
 
+        /// <summary>Gets the height of the chain.</summary>
         public abstract int Height { get; }
 
+        /// <summary>
+        /// Enumerates from start of the chain.
+        /// </summary>
+        /// <returns>An enumerable iterator.</returns>
         protected abstract IEnumerable<ChainedBlock> EnumerateFromStart();
```
Comment (Aprogiena) [inline src/NBitcoin/ChainBase.cs]: this is not for this PR, just an idea - could we get rid of this method completely? just an area to explore

File: src/NBitcoin/ChainBase.cs [hunk 4]
```diff
@@ -52,41 +75,54 @@ public ChainedBlock SetTip(ChainBase otherChain)
...
             ChainedBlock chainedHeader;
             return this.TrySetTip(header, out chainedHeader);
         }
 
+        /// <summary>
+        /// Attempts to set the tip of this chain based upon another block header.
+        /// </summary>
+        /// <param name="header">The block header to set to tip.</param>
+        /// <param name="chainedHeader">The newly chained block for the tip.</param>
+        /// <returns>Whether the tip was set successfully.</returns>
```
Comment (Aprogiena) [inline src/NBitcoin/ChainBase.cs]: it would be useful to know when such an operation can fail

File: src/NBitcoin/ChainBase.cs [hunk 5]
```diff
@@ -52,41 +75,54 @@ public ChainedBlock SetTip(ChainBase otherChain)
...
         }
 
+        /// <summary>
+        /// Whether the chain contains a block with the given hash.
+        /// </summary>
+        /// <param name="hash">The hash to search for.</param>
+        /// <returns>Whether the chain contains the block.</returns>
         public bool Contains(uint256 hash)
         {
             ChainedBlock block = GetBlock(hash);
```
Comment (Aprogiena) [inline src/NBitcoin/ChainBase.cs]: `this.` missing?

File: src/NBitcoin/ChainBase.cs [hunk 6]
```diff
@@ -52,41 +75,54 @@ public ChainedBlock SetTip(ChainBase otherChain)
...
-                throw new ArgumentNullException("blockIndex");
-
-            return this.GetBlock(blockIndex.Height) != null;
-        }
-
+        /// <summary>
+        /// Whether the given chain and this chain have the same tip.
+        /// </summary>
+        /// <param name="chain">The chain to compare this chain against.</param>
+        /// <returns>Whether the tips of the chain have the same block hash.</returns>
```
Comment (Aprogiena) [inline src/NBitcoin/ChainBase.cs]: chain`s`

File: src/NBitcoin/ChainBase.cs [hunk 7]
```diff
@@ -95,11 +131,23 @@ public bool SameTip(ChainBase chain)
             return this.Tip.HashBlock == chain.Tip.HashBlock;
         }
 
+        /// <summary>
+        /// Gets the target work required for a network and a specific block height.
+        /// </summary>
+        /// <param name="network">The network to use for work computation.</param>
+        /// <param name="height">The height of the block to get work for.</param>
+        /// <returns>The target work.</returns>
```
Comment (Aprogiena) [inline src/NBitcoin/ChainBase.cs]: `required work` here might be little better than `target work`, or `target required work`

File: src/NBitcoin/ChainBase.cs [hunk 8]
```diff
@@ -95,11 +131,23 @@ public bool SameTip(ChainBase chain)
...
             return this.GetBlock(height).GetWorkRequired(network);
         }
 
+        /// <summary>
+        /// Validates the chain on a given network.
+        /// </summary>
+        /// <param name="network">Network to validate the chain on.</param>
+        /// <param name="fullChain">Whether to validate full chain (all blocks) or just the tip.</param>
+        /// <returns>Whether the chain is valid or not.</returns>
         public bool Validate(Network network, bool fullChain = true)
```
Comment (Aprogiena) [inline src/NBitcoin/ChainBase.cs]: is this method used anywhere?
it seems like a heavy one, should not be used that much (at all?)

File: src/NBitcoin/ChainBase.cs [hunk 9]
```diff
@@ -153,6 +200,11 @@ public ChainedBlock FindFork(IEnumerable<uint256> hashes)
             return null;
         }
 
+        /// <summary>
```
Comment (Aprogiena) [inline src/NBitcoin/ChainBase.cs]: there is another version of FindFork untouched by this PR, but its comment makes no sense and need rephrasing, its summary reads as `Returns the first found block.` which is confusing

File: src/NBitcoin/ChainBase.cs [hunk 10]
```diff
@@ -161,6 +213,11 @@ public ChainedBlock FindFork(BlockLocator locator)
             return this.FindFork(locator.Blocks);
         }
 
+        /// <summary>
+        /// Enumerate chain after given block hash to genesis block.
+        /// </summary>
+        /// <param name="blockHash">Block hash to enumerate after.</param>
+        /// <returns>Enumeration of chain after given block hash.</returns>
         public IEnumerable<ChainedBlock> EnumerateAfter(uint256 blockHash)
```
Comment (Aprogiena) [inline src/NBitcoin/ChainBase.cs]: as mentioned before, these enumeration methods, if they could be checked whether they are used and whether their use is proper - they can be pretty heavy, so we should be careful

not for this PR of course

File: src/NBitcoin/ChainBase.cs [hunk 11]
```diff
@@ -171,6 +228,11 @@ public IEnumerable<ChainedBlock> EnumerateAfter(uint256 blockHash)
             return this.EnumerateAfter(block);
         }
 
+        /// <summary>
+        /// Enumerates chain from the given block to tip.
+        /// </summary>
+        /// <param name="block">Block to enumerate from</param>
+        /// <returns>Enumeration of chain from given block to tip.</returns>
         public IEnumerable<ChainedBlock> EnumerateToTip(ChainedBlock block)
```
Comment (Aprogiena) [inline src/NBitcoin/ChainBase.cs]: "Enumeration of chain" sounds confusing to me (not only this method)
i'd be more happy with
"Enumeration of chained block headers from given block to the tip."

however, these methods seem extremely dangerous
i can see a lot of them being used in tests, in which case they should be migrated to tests where it is fine to have dangerous methods
but otherwise we should refactor if any other code is using those

again, not for this PR

File: src/NBitcoin/ChainBase.cs (hunk 1 above)
Comment (dangershony) [reply src/NBitcoin/ChainBase.cs]: I would be up to renaming to `ChainedBlockHeader` even though ChainedHeader will also do the job (we don't have headers of anything else really).

Comment (mikedennis) [reply src/NBitcoin/ChainBase.cs]: Yes I agree confusion between Blocks and Chained Block Headers. I will change all the comments for now and we can decide what to do about ChainedBlock, ChainBase, ConcurrentChain longer term.

File: src/NBitcoin/ChainBase.cs (hunk 2 above)
Comment (mikedennis) [reply src/NBitcoin/ChainBase.cs]: Sounds good thanks.

File: src/NBitcoin/ChainBase.cs (hunk 4 above)
Comment (mikedennis) [reply src/NBitcoin/ChainBase.cs]: Ok sounds good.

File: src/NBitcoin/ChainBase.cs (hunk 5 above)
Comment (mikedennis) [reply src/NBitcoin/ChainBase.cs]: Yes, good catch!

File: src/NBitcoin/ChainBase.cs (hunk 7 above)
Comment (mikedennis) [reply src/NBitcoin/ChainBase.cs]: Sounds good.

File: src/NBitcoin/ChainBase.cs (hunk 11 above)
Comment (mikedennis) [reply src/NBitcoin/ChainBase.cs]: Yes I agree. I struggled with what to comment this as, I like your suggestion.

File: src/NBitcoin/ChainBase.cs (hunk 3 above)
Comment (mikedennis) [reply src/NBitcoin/ChainBase.cs]: Agreed. I find the ToEnumerable(fromTip) method confusing as well. I think it might be better to break into two separate methods rather than pass in a boolean.

File: src/NBitcoin/ChainBase.cs (hunk 8 above)
Comment (mikedennis) [reply src/NBitcoin/ChainBase.cs]: No references to it. I will delete it.

File: src/NBitcoin/ChainBase.cs (hunk 9 above)
Comment (mikedennis) [reply src/NBitcoin/ChainBase.cs]: I reworked this comment.

File: src/NBitcoin/ChainBase.cs [hunk 12]
```diff
@@ -4,15 +4,33 @@
...
 namespace NBitcoin
 {
+    /// <summary>Base implementation for chains.</summary>
     public abstract class ChainBase
     {
+        /// <summary>
+        /// Gets the chained block header given a block ID (hash).
+        /// </summary>
+        /// <param name="id">Block ID to retreive.</param>
+        /// <returns>The block.</returns>
```
Comment (Aprogiena) [inline src/NBitcoin/ChainBase.cs]: good for changing the comments to `chained block headers`, but you've forgotten some returns

File: src/NBitcoin/ChainBase.cs [hunk 13]
```diff
@@ -4,15 +4,33 @@
...
         public abstract ChainedBlock GetBlock(int height);
 
+        /// <summary>Gets the chained block header at the tip of the chain.</summary>
         public abstract ChainedBlock Tip { get; }
 
+        /// <summary>Gets the height of the chain.</summary>
         public abstract int Height { get; }
 
+        /// <summary>
+        /// Enumerates from start of the chain.
```
Comment (Aprogiena) [inline src/NBitcoin/ChainBase.cs]: Enumerates chained block headers from the start of the chain.

File: src/NBitcoin/ChainBase.cs [hunk 14]
```diff
@@ -22,14 +40,14 @@ public abstract class ChainBase
...
 
-        public virtual ChainedBlock Genesis
-        {
-            get
-            {
-                return GetBlock(0);
-            }
-        }
+        /// <summary>Gets the genesis block for the chain.</summary>
+        public virtual ChainedBlock Genesis { get { return GetBlock(0); } }
```
Comment (Aprogiena) [inline src/NBitcoin/ChainBase.cs]: `this.` missing and we can totally add it in this PR

File: src/NBitcoin/ChainBase.cs (hunk 3 above)
Comment (dangershony) [reply src/NBitcoin/ChainBase.cs]: It might also be possible to implement the IEnumerator on chained block itself, but only if it will be really beneficial.

File: src/NBitcoin/ChainBase.cs [hunk 15]
```diff
@@ -52,41 +75,54 @@ public ChainedBlock SetTip(ChainBase otherChain)
...
             ChainedBlock chainedHeader;
             return this.TrySetTip(header, out chainedHeader);
         }
 
+        /// <summary>
+        /// Attempts to set the tip of this chain based upon another block header.
+        /// </summary>
+        /// <param name="header">The block header to set to tip.</param>
+        /// <param name="chainedHeader">The newly chained block header for the tip.</param>
+        /// <returns>Whether the tip was set successfully. Will return <c>false</c> if the block header previous block link cannot be found in the current chain.</returns>
```
Comment (Aprogiena) [inline src/NBitcoin/ChainBase.cs]: hard to read the second sentence :-)
how about
`<returns>Whether the tip was set successfully. The method fails (and returns <c>false</c>) if the <paramref name="header"/>'s link to a previous header does not point to any block in the current chain.</returns>`

File: src/NBitcoin/ChainBase.cs [hunk 16]
```diff
@@ -161,6 +189,11 @@ public ChainedBlock FindFork(BlockLocator locator)
             return this.FindFork(locator.Blocks);
         }
 
+        /// <summary>
+        /// Enumerate chain after given block hash to genesis block.
+        /// </summary>
+        /// <param name="blockHash">Block hash to enumerate after.</param>
+        /// <returns>Enumeration of chained block headers after given block hash.</returns>
```
Comment (dangershony) [inline src/NBitcoin/ChainBase.cs]: This feels to me like it could be an extension method, perhaps it might be a good idea to see what other methods can be pulled as extensions (not this PR)

File: src/NBitcoin/ChainBase.cs [hunk 17]
```diff
@@ -95,35 +131,22 @@ public bool SameTip(ChainBase chain)
...
-
-
         /// <summary>
-        /// Returns the first common block between two chains.
+        /// Returns the first common chained block header between two chains.
         /// </summary>
         /// <param name="chain">The other chain.</param>
-        /// <returns>First common block or <c>null</c>.</returns>
+        /// <returns>First common chained block header or <c>null</c>.</returns>
         public ChainedBlock FindFork(ChainBase chain)
```
Comment (Aprogiena) [inline src/NBitcoin/ChainBase.cs]: this method is implemented terribly, it goes through whole chain!? fortunately it is used only in tests - it should be moved to tests

please do it as a part of this PR because this method is terrible

File: src/NBitcoin/ChainBase.cs [hunk 18]
```diff
@@ -133,10 +156,10 @@ public ChainedBlock FindFork(ChainBase chain)
         }
 
         /// <summary>
-        /// Returns the first found block.
+        /// Returns the first chained block header that exists in the chain from the list of block hashes.
         /// </summary>
         /// <param name="hashes">Hash to search for.</param>
-        /// <returns>First found block or <c>null</c>.</returns>
+        /// <returns>First found chained block header or <c>null</c>.</returns>
```
Comment (Aprogiena) [inline src/NBitcoin/ChainBase.cs]: or null ... when null? null if ...

File: src/NBitcoin/ChainBase.cs [hunk 19]
```diff
@@ -179,6 +217,11 @@ public IEnumerable<ChainedBlock> EnumerateToTip(ChainedBlock block)
             return this.EnumerateToTip(block.HashBlock);
         }
 
+        /// <summary>
+        /// Enumerates chain from given block hash to tip.
```
Comment (Aprogiena) [inline src/NBitcoin/ChainBase.cs]: Enumerates chained block headers

File: src/NBitcoin/ChainBase.cs [hunk 20]
```diff
@@ -171,6 +204,11 @@ public IEnumerable<ChainedBlock> EnumerateAfter(uint256 blockHash)
             return this.EnumerateAfter(block);
         }
 
+        /// <summary>
+        /// Enumerates chain from the given chained block header to tip.
+        /// </summary>
+        /// <param name="block">Chained block header to enumerate from</param>
```
Comment (Aprogiena) [inline src/NBitcoin/ChainBase.cs]: missing dot

File: src/NBitcoin/ChainBase.cs [hunk 21]
```diff
@@ -171,6 +204,11 @@ public IEnumerable<ChainedBlock> EnumerateAfter(uint256 blockHash)
             return this.EnumerateAfter(block);
         }
 
+        /// <summary>
+        /// Enumerates chain from the given chained block header to tip.
```
Comment (Aprogiena) [inline src/NBitcoin/ChainBase.cs]: Enumerates chained block headers

File: src/NBitcoin/ChainBase.cs [hunk 22]
```diff
@@ -191,6 +234,11 @@ public IEnumerable<ChainedBlock> EnumerateToTip(uint256 blockHash)
                 yield return chainedBlock;
         }
 
+        /// <summary>
+        /// Enumerates chain after the given chained block header to genesis block.
```
Comment (Aprogiena) [inline src/NBitcoin/ChainBase.cs]: chained block headers
